---
title: "Variant Calling and QC"
author: "Paul Theodor Pyl"
date: "12 Jun 2015"
output:
  ioslides_presentation:
    css: style.css
runtime: shiny
---

# Basics of sequencing-based Variant Calling

## Single Nucleotide Variants (SNVs)

![SNVsOverview][snvOV]

## Small Insertions and Deletions (InDels)

![InDelOverview][InDelOV]

## The Data (SNV example)

![SNVData][snvData]

## The Data (SNV example) 

Position  |  0 |  1 |  2 |  3 | 4 | 5 |  6 |  7 |  8 |  9 | 10 | 11 |
---------:| --:| --:| --:| --:|--:|--:| --:| --:| --:| --:| --:| --:|
A         |  0 |  6 |  0 |  **6** |  0 |  0 |  0 |  0 |  5 |  0 |  0 |  0 |
C         |  *1* |  0 |  0 |  **4** |  0 |  0 |  0 |  0 |  0 |  4 |  0 |  0 |
G         |  2 |  0 |  0 |  0 | 10 |  0 |  **8** |  8 |  0 |  0 |  2 |  0 |
T         |  0 |  0 |  9 |  0 |  0 | 10 |  0 |  0 |  *1* |  0 |  0 |  1 |

## The Data (Cohort Example)
<div style="width: 70%; margin: auto;">
![cohortTally][cohortTally]
</div>

## Variant-Calling Software
- [The Genome Analysis Toolkit](https://www.broadinstitute.org/gatk/)
- [Samtools](http://www.htslib.org/doc/samtools-1.2.html)
- Bioconductor Packages:
    - [VariantTools](http://bioconductor.org/packages/release/bioc/html/VariantTools.html)
    - [h5vc](http://bioconductor.org/packages/release/bioc/html/h5vc.html)

![h5vc][h5vc]

## Certainty and Sequencing Depth
In Diploid Organisms heterozygous variants should be observed with an average frequency of $0.5$, i.e. $50\%$ of the reads overlapping the position should show the variant allele.

## Certainty and Sequencing Depth

```{r, echo=FALSE, eval = TRUE}
inputPanel(
  sliderInput("s", label = "Sequencing Depth:",
              min = 10, max = 100, step = 10, value = 20),
  
  sliderInput("af", label = "Expected allele frequency:",
              min = 0.1, max = 1, value = 0.5, step = 0.1),
  
  checkboxInput("count", label = "Show Allele Frequency", value = TRUE)
)

renderPlot({
  if(input$count){
    d <- input$s
  }else{
    d <- 1
  }
  plot(
    seq(0,input$s)/d,
    dbinom(x = seq(0, input$s), size = input$s, prob = input$af),
    type = "h", xlab = "Allele Frequency (#SNV / #Reference)", ylab = "Binomial Model Density",
    main = paste( "Example with sequencing depth", input$s )
    )
})
```

## Somatic Variants vs. Germline Variants

When dealing with paired cancer samples we ask:

* What are the variants that the person is carrying (germline variants)?
* What are the variants that the cancer is carrying (somatic variants)?
* Which somatic variants are causal to the cancer development / progression (driver variants)?
* Which somatic variants are a byproduct of genomic instability (passenger variants)?

## You've gotten some variants, now what? 

* Visualising them helps to spot weird patterns
* Annotate them to gain extra information, e.g. population frequencies, SIFT scores, affected genes and type of mutation (missense, nonsense, stop gain, ...)

## Telling the good from the bad

![GoodSNV][good]

## Telling the good from the bad

![BadSNV][bad]

## Telling the good from the bad
<div style="width: 80%; margin: auto;">
![nice][nice]
</div>
## Telling the good from the bad
<div style="width: 80%; margin: auto;">
![notSoNice][notSoNice]
</div>
## Telling the good from the bad
<div style="width: 70%; margin: auto;">
![horrible][horrible]
</div>

# The VCF File Format
## Variant File Formats {.smaller}
```{r, message=FALSE}
require(VariantAnnotation)
vcf <- readVcf(file = "1000g.subset.vcf", genome = "hg19")
vcf
```

## Variant File Formats {.smaller}
```{r}
head(rowRanges(vcf))
```

## Variant File Formats {.smaller}
```{r}
geno(vcf)$GT[1:3,1089:1093]
```

```{r}
info(vcf)[1:3,grep(x = colnames(info(vcf)), pattern = "AF")]
```

# Annotating Variants

## Ensembl Variant Effect Predictor

![vep][vep]

## Ensembl Variant Effect Predictor

```{r, eval = FALSE}
source("http://bioconductor.org/biocLite.R")
biocLite("ensemblVEP")
```

Interesting Annotations:

- Affected Genes
- Missense, Nonsense, splice site, ...
- SIFT and PolyPhen Scores (conservation)
- Populations allele frequencies and dbSNP / COSMIC IDs

## Ensembl Variant Effect Predictor - Report
<div style="width: 70%; margin: auto;">
![vepReport][vepReport]
</div>
## Ensembl Variant Effect Predictor - Results {.smaller}
```{r, echo=TRUE}
vep <- read.table("VEP.Results.txt", header = TRUE, stringsAsFactors = FALSE)
colnames(vep)
```

## Ensembl Variant Effect Predictor - Results {.smaller}
```{r, echo=FALSE}
x <- unlist(subset(vep, SIFT != "-"))
x[x!="-"]
```

## Ensembl Variant Effect Predictor - Results {.smaller}

```{r}
x <- subset(vep, GMAF != "-")
x[,c("Location", "Allele", "SYMBOL", grep("MAF", colnames(x), value = TRUE))]
```

## Ensembl VEP - Impact Scores
We use the SIFT and PolyPhen scores to asses how damaging a variant is.

- Based on homology, i.e. only valid for coding variants
- For a given position we model the likelihood of a homologous position being mutated
- If a position can be mutated a lot in homologous genes, it is less likely to be important for the gene function

## Ensembl VEP - Known Variants
dbSNP:

- for population variants (Single Nucleotide **Polymorphisms**)

COSMIC:

- for somatic variants reported in cancer studies

## The End

## Plotting SNVs - Comparative Example

```{r, echo=FALSE, message=FALSE, eval=TRUE,}
require(h5vc)
tallyFile <- system.file( "extdata", "example.tally.hfs5", package = "h5vcData" )
sampleData <- getSampleData( tallyFile, "/ExampleStudy/16" )
position <- 29983015

windowsize <- 2000
vars <- h5dapply( # Calling Variants
  filename = tallyFile,
  group = "/ExampleStudy/16",
  blocksize = 500,
  FUN = callVariantsPaired,
  sampledata = sampleData,
  cl = vcConfParams(returnDataPoints=TRUE),
  names = c("Coverages", "Counts", "Reference", "Deletions"),
  range = c(position - windowsize, position + windowsize)
)
vars <- do.call( rbind, vars ) # merge the results from all blocks by row

windowsize = 30
require(GenomicRanges)
ranges = GRanges("16", ranges = IRanges(start = vars$Start - windowsize, end = vars$End + windowsize), strand = "*")

dataList <- h5dapply(
  filename = tallyFile,
  group = "/ExampleStudy",
  names = c("Coverages", "Counts", "Deletions", "Reference"),
  range = ranges
)

pll <- mismatchPlot(
  data = dataList[["16"]],
  sampledata = sampleData
)
print(pll)
```



[snvOV]: Illustrations/SNVOverview.png "Overview of SNV types"
[snvData]: Illustrations/SNVdata.png "Overview of WGS/WEX SNV Data"
[InDelOV]: Illustrations/InDelOverview.png "Overview of InDel types"
[good]: Illustrations/CSMD3.png "A Good Variant"
[bad]: Illustrations/PABPC1.png "A Bad Variant"
[vep]: Illustrations/vep.png "the VEP"
[vepReport]: Illustrations/vepReport.png "the VEP Report"
[cohortTally]: Illustrations/cohortTally.png "A Tally of a cohort"
[nice]: Illustrations/nice.tiff "A Good Variant"
[notSoNice]: Illustrations/not.so.nice.tiff "A Bad Variant"
[horrible]: Illustrations/patently.horrible.png "A Very Bad Variant"
[h5vc]: Illustrations/h5vc.png "The h5vc Paper"