---
title: "Lab: Introduction to _R_ and _Bioconductor_"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  % \VignetteIndexEntry{Lab: Introduction to R and Bioconductor}
  % \VignetteEngine{knitr::rmarkdown}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

```{r setup, echo=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

Authors: Martin Morgan (<a
  href="mailto:mtmorgan@fhcrc.org">mtmorgan@fhcrc.org</a>), Sonali
  Arora (<a
  href="mailto:sarora@fredhutch.org">sarora@fredhutch.org</a><br />
Date: 15 June, 2015

# R

## Refresher: working with data

### Case study: ALL phenotypic data

This case study servers as a refresher basic input and manipulation of
data.

Input a file that contains ALL (acute lymphoblastic leukemia) patient
information

```{r echo=TRUE, eval=FALSE}
fname <- file.choose()   ## "ALLphenoData.tsv"
stopifnot(file.exists(fname))
pdata <- read.delim(fname)
```
```{r echo=FALSE}
fname <- system.file(package="CSAMA15Morgan", "extdata",
    "ALLphenoData.tsv")
stopifnot(file.exists(fname))
pdata <- read.delim(fname)
```

Check out the help page `?read.delim` for input options, and explore
basic properties of the object you've created, for instance...

```{r ALL-properties}
class(pdata)
colnames(pdata)
dim(pdata)
head(pdata)
summary(pdata$sex)
summary(pdata$cyto.normal)
```

Remind yourselves about various ways to subset and access columns of a
data.frame

```{r ALL-subset}
pdata[1:5, 3:4]
pdata[1:5, ]
head(pdata[, 3:5])
tail(pdata[, 3:5], 3)
head(pdata$age)
head(pdata$sex)
head(pdata[pdata$age > 21,])
```

It seems from below that there are 17 females over 40 in the data set,
but when sub-setting `pdata` to contain just those individuals 19 rows
are selected. Why? What can we do to correct this?

```{r ALL-subset-NA}
idx <- pdata$sex == "F" & pdata$age > 40
table(idx)
dim(pdata[idx,])
```

Use the `mol.biol` column to subset the data to contain just
individuals with 'BCR/ABL' or 'NEG', e.g.,

```{r ALL-BCR/ABL-subset}
bcrabl <- pdata[pdata$mol.biol %in% c("BCR/ABL", "NEG"),]
```

The `mol.biol` column is a factor, and retains all levels even after
subsetting. How might you drop the unused factor levels?

```{r ALL-BCR/ABL-drop-unused}
bcrabl$mol.biol <- factor(bcrabl$mol.biol)
```

The `BT` column is a factor describing B- and T-cell subtypes

```{r ALL-BT}
levels(bcrabl$BT)
```

How might one collapse B1, B2, ... to a single type B, and likewise for T1, T2, ..., so there are only two subtypes, B and T

```{r ALL-BT-recode}
table(bcrabl$BT)
levels(bcrabl$BT) <- substring(levels(bcrabl$BT), 1, 1)
table(bcrabl$BT)
```

Use `xtabs()` (cross-tabulation) to count the number of samples with
B- and T-cell types in each of the BCR/ABL and NEG groups

```{r ALL-BCR/ABL-BT}
xtabs(~ BT + mol.biol, bcrabl)
```

Use `aggregate()` to calculate the average age of males and females in
the BCR/ABL and NEG treatment groups.

```{r ALL-aggregate}
aggregate(age ~ mol.biol + sex, bcrabl, mean)
```

Use `t.test()` to compare the age of individuals in the BCR/ABL versus
NEG groups; visualize the results using `boxplot()`. In both cases,
use the `formula` interface. Consult the help page `?t.test` and re-do
the test assuming that variance of ages in the two groups is
identical. What parts of the test output change?

```{r ALL-age}
t.test(age ~ mol.biol, bcrabl)
boxplot(age ~ mol.biol, bcrabl)
```

### Case study: weighty matters

This case study is a second walk through basic data manipulation and
visualization skills.  We use data from the US Center for Disease
Control's Behavioral Risk Factor Surveillance System ([BRFSS][])
annual survey. Check out the web page for a little more
information. We are using a small subset of this data, including a
random sample of 10000 observations from each of 1990 and 2010.

Input the data using `read.csv()`, creating a variable `brfss` to hold
it.  Use `file.choose()` to locate the data file BRFSS-subset.csv

```{r echo=TRUE, eval=FALSE}
fname <- file.choose()   ## BRFSS-subset.csv
stopifnot(file.exists(fname))
brfss <- read.csv(fname)
```
```{r echo=FALSE}
fname <- system.file(package="CSAMA15Morgan", "extdata",
    "BRFSS-subset.csv")
stopifnot(file.exists(fname))
brfss <- read.csv(fname)
```

#### Exercises: base plotting functions

1. Explore the data using `class()`, `dim()`, `head()`, `summary()`,
   etc. Use `xtabs()` to summarize the number of males and females in
   the study, in each of the two years.

2. Use `aggregate()` to summarize the average weight in each sex and
   year.
  
3. Create a scatterplot showing the relationship between the square
   root of weight and height, using the `plot()` function and the
   `main` argument to annotate the plot. Note the transformed
   Y-axis. Experiment with different plotting symbols (try the command
   `example(points)` to view different points).

   ```{r brfss-simple-plot}
   plot(sqrt(Weight) ~ Height, brfss, main="All Years, Both Sexes")
   ```
   
4. Color the female and male points differently. To do this, use the
   `col` argument to `plot()`. Provide as a value to that argument a
   vector of colors, subset by `brfss$Sex`.

5. Create a subset of the data containing only observations from
   2010.

   ```{r brfss-subset}
   brfss2010 <- brfss[brfss$Year == "2010", ]
   ```

6. Create the figure below (two panels in a single figure). Do this by
   using the `par()` function with the `mfcol` argument before calling
   `plot()`. You'll need to create two more subsets of data, perhaps
   when you are providing the data to the function `plot`.

   ```{r brfss-pair-plot}
   opar <- par(mfcol=c(1, 2))
   plot(sqrt(Weight) ~ Height, brfss2010[brfss2010$Sex == "Female", ],
        main="2010, Female")
   plot(sqrt(Weight) ~ Height, brfss2010[brfss2010$Sex == "Male", ],
        main="2010, Male")
   par(mfcol=c(1, 1))
   ```

7. Plotting large numbers of points means that they are often
   over-plotted, potentially obscuring important patterns. Experiment
   with arguments to `plot()` to address over-plotting, e.g.,
   `pch='.'` or `alpha=.4`. Try using the `smoothScatter()` function
   (the data have to be presented as `x` and `y`, rather than as a
   formula). Try adding the [hexbin][] library to your R session
   (using `library()`) and creating a `hexbinplot()`.

#### Exercises: Trellis graphics and the _lattice_ package

R has a number of additional plotting facilities, both as part of the
'base' distribution and user-contributed packages. The _lattice_
package adopts a particular philosophy about the presentation of data,
and can be a very effective visualization tool.

1.  Use `library()` to load the _lattice_ package.

    ```{r lattice}
    library(lattice)
    ```

2. Create the following figure using the `xyplot()` function with a
   formula and the `brfss2010` data. The formula is `sqrt(Weight) ~
   Height | Sex`, which can be read as `square root of Weight as a
   function of Height, conditioned on Sex'.

   ```{r lattice-conditioning}
   xyplot(sqrt(Weight) ~ Height | Sex, brfss2010)
   ```

3. Add a background grid and a regression line to each panel using the
   argument `type=c('p', 'g', 'r')`; change the width (`lwd`) and
   color (`col.line`) of the regression line. For some hints on other
   arguments, see the help pages `?xyplot` (for overall structure of
   the plot) and `?panel.xyplot` (for how each 'panel' of data is
   plotted).
  
4. Create the following figure. Use the `densityplot()` function with
   the formula `~ sqrt(Weight)`. The `group=Sex` function argument
   creates separate lines for each sex. Be sure to use
   `plot.points=FALSE` to avoid a `rug' of points at the base of the
   figure. Can you add a key (legend)?
   
   ```{r lattice-density}
   densityplot(~sqrt(Weight), brfss2010, group=Sex, plot.points=FALSE)
   ```

5. Create the figure below using the `bwplot` function. The formula
   requires that `Year` be coerced to a `factor`, `factor(Year)`.

    ```{r lattice-bwplot}
    bwplot(sqrt(Weight) ~ factor(Year) | Sex, brfss)
    ```
  
6.  Create the Figure below, a _violin_ plot, using `bwplot()` and the
    `panel` argument set to `panel.violin`. from `?bwplot` we learn
    that `panel` is a function that determines how each panel is
    drawn; the details for controling the violin panel plot are
    described on the help page `?panel.violin`.

    ```{r lattice-violin}
    bwplot(sqrt(Weight) ~ factor(Year) | Sex, brfss, panel=panel.violin)
    ```

7. (Advanced) We can write our own `panel` argument to _lattice_
   functions to influence how each panel is displayed. Here we add a
   point at the median age and weight.

    ```{r lattice-panel}
    xyplot(sqrt(Weight) ~ Height|Sex, brfss2010,
        panel = function(x, y, ...) {
            panel.xyplot(x, y, ...)
            panel.points(median(x, na.rm=TRUE), median(y, na.rm=TRUE), 
                cex=2, pch=20, col="red")
        },
        type=c("p", "g", "r"), lwd=2, col.line="red", xlim=c(120, 210))
    ```

# _Bioconductor_

## Making progress: classes, methods, and packages

This section focuses on classes, methods, and packages, with the goal
being to learn to navigate the help system and interactive discovery
facilities.

### Motivation

Sequence analysis is specialized
- Large data needs to be processed in a memory- and time-efficient manner
- Specific algorithms have been developed for the unique
  characteristics of sequence data

Additional considerations
- Re-use of existing, tested code is easier to do and less error-prone
  than re-inventing the wheel.
- Interoperability between packages is easier when the packages share
  similar data structures.

Solution: use well-defined _classes_ to represent complex data;
_methods_ operate on the classes to perform useful functions.  Classes
and methods are placed together and distributed as _packages_ so that
we can all benefit from the hard work and tested code of others.

### Case study: IRanges and GRanges

The [IRanges][] package defines an important class for specifying
integer ranges, e.g.,
```{r iranges}
library(IRanges)
ir <- IRanges(start=c(10, 20, 30), width=5)
ir
```

There are many interesting operations to be performed on ranges, e.g,
`flank()` identifies adjacent ranges
```{r iranges-flank}
flank(ir, 3)
```

Consult the help page for flank, `?flank`, and explore other
range-based operations.

The `IRanges` class is part of a class hierarchy. To see this, ask R for
the class of `ir`, and for the class definition of the `IRanges` class
```{r iranges-class}
class(ir)
getClassDef(class(ir))
```

Notice that `IRanges` extends the `Ranges` class. Now try entering
`?"flank,<tab>`, where `<tab>` means to press the tab key to ask for
tab completion (may not be necessary in Rstudio). You can see that
there are help pages for several different classes. Tab-complete to

```{r iranges-flank-method, eval=FALSE}
?"flank,Ranges-method" 
```

and verify that you're at the page that describes the method relevant
to an `IRanges` instance.

The [GenomicRanges][] package extends the notion of ranges to include
features relevant to application of ranges in sequence analysis,
particularly the ability to associate a range with a sequence name
(e.g., chromosome) and a strand. Create a `GRanges` instance based on
our `IRanges` instance, as follows
```{r granges}
library(GenomicRanges)
gr <- GRanges(c("chr1", "chr1", "chr2"), ir, strand=c("+", "-", "+"))
gr
```

The notion of flanking sequence has a more nuanced meaning in
biology. In particular we might expect that flanking sequence on the
`+` strand would precede the range, but on the minus strand would
follow it. Verify that `flank` applied to a `GRanges` object has this
behavior.
```{r granges-flank}
flank(gr, 3)
```

Discover what classes `GRanges` extends, find the help page
documenting the behavior of `flank` when applied to a GRanges object,
and verify that the help page documents the behavior we just observed.
```{r granges-class}
class(gr)
getClassDef(class(gr))
```
```{r granges-flank-method, eval=FALSE}
?"flank,GenomicRanges-method"
```
Notice that the available `flank()` methods have been augmented by the
methods defined in the _GenomicRanges_ package.

It seems like there might be a number of helpful methods available for
working with genomic ranges; we can discover some of these from the
command line, indicating that the methods should be on the current
`search()` path

```{r granges-methods, eval=FALSE}
showMethods(class="GRanges", where=search())
```

Use `help()` to list the help pages in the `GenomicRanges` package,
and `vignettes()` to view and access available vignettes; these are
also available in the Rstudio 'Help' tab.
```{r granges-man-and-vignettes, eval=FALSE}
help(package="GenomicRanges")
vignette(package="GenomicRanges")
vignette(package="GenomicRanges", "GenomicRangesHOWTOs")
```

## Annotations

(see Rnw file)

### Genes

### Genomes

### _AnnotationHub_

## A sequence analysis package tour

This very open-ended topic points to some of the most prominent
Bioconductor packages for sequence analysis. Use the opportunity in
this lab to explore the package vignettes and help pages highlighted
below; many of the material will be covered in greater detail in
subsequent labs and lectures.

Basics 

- Bioconductor packages are listed on the [biocViews][] page. Each
  package has 'biocViews' (tags from a controlled vocabulary)
  associated with it; these can be searched to identify appropriately
  tagged packages, as can the package title and author.
- Each package has a 'landing page', e.g., for
  [GenomicRanges][]. Visit this landing page, and note the
  description, authors, and installation instructions. Packages are
  often written up in the scientific literature, and if available the
  corresponding citation is present on the landing page. Also on the
  landing page are links to the vignettes and reference manual and, at
  the bottom, an indication of cross-platform availability and
  download statistics.
- A package needs to be installed once, using the instructions on the
  landing page. Once installed, the package can be loaded into an R
  session and the help system queried interactively, as outlined
  above:

```{r require}
library(GenomicRanges)
```

```{r help, eval=FALSE}
help(package="GenomicRanges")
vignette(package="GenomicRanges")
vignette(package="GenomicRanges", "GenomicRangesHOWTOs")
?GRanges
```

Domain-specific analysis -- explore the landing pages, vignettes, and
reference manuals of two or three of the following packages.

- Important packages for analysis of differential expression include
  [edgeR][] and [DESeq2][]; both have excellent vignettes for
  exploration. Additional research methods embodied in Bioconductor
  packages can be discovered by visiting the [biocViews][] web page,
  searching for the 'DifferentialExpression' view term, and narrowing
  the selection by searching for 'RNA seq' and similar.
- Popular ChIP-seq packages include [DiffBind][] for comparison of
  peaks across samples, [ChIPQC][] for quality assessment, and
  [ChIPpeakAnno][] for annotating results (e.g., discovering nearby
  genes). What other ChIP-seq packages are listed on the [biocViews][]
  page?
- Working with called variants (VCF files) is facilitated by packages
  such as [VariantAnnotation][], [VariantFiltering][], and
  [ensemblVEP][]; packages for calling variants include, e.g.,
  [h5vc][] and [VariantTools][].
- Several packages identify copy number variants from sequence data,
  including [cn.mops][]; from the [biocViews][] page, what other copy
  number packages are available? The [CNTools][] package provides some
  useful facilities for comparison of segments across samples.
- Microbiome and metagenomic analysis is facilitated by packages such
  as [phyloseq][] and [metagenomeSeq][].
- Metabolomics, chemoinformatics, image analysis, and many other
  high-throughput analysis domains are also represented in
  Bioconductor; explore these via biocViews and title searches.
  
Working with sequences, alignments, common web file formats, and raw
data; these packages rely very heavily on the [IRanges][] /
[GenomicRanges][] infrastructure that we will encounter later in the
course.

- The [Biostrings][] package is used to represent DNA and other
  sequences, with many convenient sequence-related functions. Check
  out the functions documented on the help page `?consensusMatrix`,
  for instance. Also check out the [BSgenome][] package for working
  with whole genome sequences, e.g., `?"getSeq,BSgenome-method"`
- The [GenomicAlignments][] package is used to input reads aligned to
  a reference genome. See for instance the `?readGAlignments` help
  page and `vigentte(package="GenomicAlignments",
  "summarizeOverlaps")`
- [rtracklayer][]'s `import` and `export` functions can read in many
  common file types, e.g., BED, WIG, GTF, ..., in addition to querying
  and navigating the UCSC genome browser. Check out the `?import` page
  for basic usage.
- The [ShortRead][] and [Rsamtools][] packages can be used for
  lower-level access to FASTQ and BAM files, respectively. Explore the
  [ShortRead vignette](http://bioconductor.org/packages/release/bioc/vignettes/ShortRead/inst/doc/Overview.pdf)
  and Scalable Genomics labs to see approaches to effectively
  processing the large files.

Annotation: Bioconductor provides extensive access to 'annotation'
resources (see the [AnnotationData][] biocViews hierarchy); these are
covered in greater detail in Thursday's lab, but some interesting
examples to explore during this lab include:

- [biomaRt][], [PSICQUIC][], [KEGGREST][] and other packages for
  querying on-line resources; each of these have informative vignettes.
- [AnnotationDbi][] is a cornerstone of the
  [Annotation Data][AnnotationData] packages provided by Bioconductor.
  - **org** packages (e.g., [org.Hs.eg.db][]) contain maps between
    different gene identifiers, e.g., ENTREZ and SYMBOL. The basic
    interface to these packages is described on the help page `?select`
  - **TxDb** packages (e.g., [TxDb.Hsapiens.UCSC.hg19.knownGene][])
    contain gene models (exon coordinates, exon / transcript
    relationships, etc) derived from common sources such as the hg19
    knownGene track of the UCSC genome browser. These packages can be
    queried, e.g., as described on the `?exonsBy` page to retrieve all
    exons grouped by gene or transcript.
  - **BSgenome** packages (e.g., [BSgenome.Hsapiens.UCSC.hg19][])
    contain whole genomes of model organisms.
- [VariantAnnotation][] and [ensemblVEP][] provide access to sequence
  annotation facilities, e.g., to identify coding variants; see the
  [Introduction to VariantAnnotation](http://bioconductor.org/packages/release/bioc/vignettes/ShortRead/inst/doc/Overview.pdf)
  vignette for a brief introduction; we'll re-visit this during the
  Thursday lab.
- Take a quick look (we'll do more of this in Thursday's lab) at the
  [annotation work flow](http://bioconductor.org/help/workflows/annotation/annotation/)
  on the Bioconductor web site.



[BRFSS]: http://www.cdc.gov/brfss/

[biocViews]: http://bioconductor.org/packages/BiocViews.html#___Software
[AnnotationData]: http://bioconductor.org/packages/BiocViews.html#___AnnotationData

[aprof]: http://cran.r-project.org/web/packages/aprof/index.html
[hexbin]: http://cran.r-project.org/web/packages/hexbin/index.html
[lineprof]: https://github.com/hadley/lineprof
[microbenchmark]: http://cran.r-project.org/web/packages/microbenchmark/index.html


[AnnotationDbi]: http://bioconductor.org/packages/AnnotationDbi.html
[BSgenome]: http://bioconductor.org/packages/BSgenome.html
[Biostrings]: http://bioconductor.org/packages/Biostrings.html
[CNTools]: http://bioconductor.org/packages/CNTools.html
[ChIPQC]: http://bioconductor.org/packages/ChIPQC.html
[ChIPpeakAnno]: http://bioconductor.org/packages/ChIPpeakAnno.html
[DESeq2]: http://bioconductor.org/packages/DESeq2.html
[DiffBind]: http://bioconductor.org/packages/DiffBind.html
[GenomicAlignments]: http://bioconductor.org/packages/GenomicAlignments.html
[GenomicRanges]: http://bioconductor.org/packages/GenomicRanges.html
[IRanges]: http://bioconductor.org/packages/IRanges.html
[KEGGREST]: http://bioconductor.org/packages/KEGGREST.html
[PSICQUIC]: http://bioconductor.org/packages/PSICQUIC.html
[Rsamtools]: http://bioconductor.org/packages/Rsamtools.html
[ShortRead]: http://bioconductor.org/packages/ShortRead.html
[VariantAnnotation]: http://bioconductor.org/packages/VariantAnnotation.html
[VariantFiltering]: http://bioconductor.org/packages/VariantFiltering.html
[VariantTools]: http://bioconductor.org/packages/VariantTools.html
[biomaRt]: http://bioconductor.org/packages/biomaRt.html
[cn.mops]: http://bioconductor.org/packages/cn.mops.html
[h5vc]: http://bioconductor.org/packages/h5vc.html
[edgeR]: http://bioconductor.org/packages/edgeR.html
[ensemblVEP]: http://bioconductor.org/packages/ensemblVEP.html
[limma]: http://bioconductor.org/packages/limma.html
[metagenomeSeq]: http://bioconductor.org/packages/metagenomeSeq.html
[phyloseq]: http://bioconductor.org/packages/phyloseq.html
[snpStats]: http://bioconductor.org/packages/snpStats.html

[org.Hs.eg.db]: http://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html
[TxDb.Hsapiens.UCSC.hg19.knownGene]: http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html
[BSgenome.Hsapiens.UCSC.hg19]: http://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg19.html
